## Promise类
- Promise类的参数是一个执行器executor,会立即执行
- executor中有两个参数，分别是resolve、reject两个函数，成功会执行resolve,失败会执行reject  抛出错误时同样会执行reject
- 在外面调用resolve和reject时，传入的参数，会赋值到Promise类（或者说是实例上）this.value = value  并且改变state
- 三个状态：pending、fulfilled、rejected  只有在等待态才能变成成功态或失败态，fulfilled和rejected是不能相互转换的
- 当调用resolve或reject是异步时，需要将多个then中的onfulfilled和onrejected分别放入两个数组，在执行resolve或reject时，挨个执行

let promise = new Promise((resolve,reject) => {
    console.log('会立即执行')
    resolve('success')
    reject('fail')
})


## promise实例上的then方法
- then方法中有两个参数：onfulfilled,onrejected
- 为实现链式调用，需要在then中返回一个new Promise()
- 根据this.status的值，去分别执行 onfulfilled,onrejected，其中return出去的111给了let x = onfulfilled(this.value)中的x

// 执行：promise.then(()=> { return 111},err => {}).then(data => {},err => {})
// 原理：promise.then = function(onfulfilled,onrejected){}


## 判断 promise.then(()=> { return 111}）中return出的值的类型
- 判断x的值是promise还是一个普通值
-- 如果是一个普通值，直接resolve(),会走下一个then中的onfulfilled成功回调
-- 如果是x不是null,并且它是一个函数或对象，就可以初步判断是一个promise
--- 看看x有没有then方法，并且then是不是一个函数，如果是，就可以确定它是promise
--- 调用这个then方法，并且根据x(x是promise)中是resolve或者reject,来执行then中的onfulfilled,onrejected，并且将结果传给下一个then作为参数
// 执行：promise.then(()=> { return new Promise((resolve,reject)=> { // 结果决定下一个then走onfulfilled或onrejected })},err => {}).then(data => {},err => {})

## return问题
- return data 给了下一个then作为参数
- promise1、promise2肯定是一个新的promise实例
let p = new Promise((resolve,reject) => {
    resolve('success')
})
let promise1 = p.then(data => {
    // return data
    return new Promise(resolve => { // return出去的值都是传给了下一个then方法中的参数
        resolve(123)
    })
})
let promise2 = promise1.then(str => {
    console.log(str)
})